doc: ieee802154: improved documentation structure, OT cross-reference and documentation details

fgrandel opened the pull request:
This PR intends to improve the overall IEEE 802.15.4 documentation and resolve ambivalence re it's stability:
* Adds the IEEE 802.15.4 documentation to the API overview and starts to version three separate IEEE 802.15.4 sub-APIs.
* Divides the IEEE 802.15.4 documentation into three parts: l2, driver and mgmt and documents the relevance of these parts to different audiences
* Adds cross references between Thread/OpenThread and native L2 as both stacks share the same driver API
* Improves type-level grouping and documentation and hides types of no public interest

Currently all L2/L1s seem to be subsumed under "network" but IMO that doesn't make sense as it does not allow us to address the different versioning requirements of IEEE 802.15.4 sub-APIs nor does it allow us to differentiate the stability of Ethernet, Bluetooth, CAN and IEEE 802.15.4.

This PR therefore proposes to version the IEEE 802.15.4 subsystem at three independent levels (all of which I consider UNSTABLE for now):
* **L2**: internal API needed to interface with the net stack, the IEEE 802.15.4 specific part (ieee802154_context/pkt) very unstable due to active protocol development, audience: **subsystem maintainers only**
* **drivers**: internal API, currently very unstable, requires a lot of change, audience: **driver maintainers only**
* **net_mgmt**: API exposed to applications, mostly only additions are required, could be stable but I wouldn't do that, yet, audience: **application developers**

This allows us to address the different requirements of the three sub-APIs re versioning and change management:
* Breaking net_mgmt really causes headaches to existing applications - so that one we should keep as stable as possible.
* The driver API will break out-of-tree drivers, but so far I haven't heard of a single existing instance of one from the community.
* The L2 API is not a problem at all AFAICS as it is used in-tree only.

For all practical purposes, @rlubos, @jukkar  and I considered the existing user base of those APIs small and the changes required for its transformation many so that we agreed to treat it de facto as unstable over the last three releases (which I'm proposing to formalize now).

This strategy seems to have worked out well. AFAICS we didn't get a single complaint and a lot of positive feedback for that approach from the community.

Commit:
doc: ieee802154: improved API documentation structure

Improves the API documentation including it's structure and integration
into the Zephyr documentation.

Previously the API status of the IEEE 802.15.4 API (and its sub-APIs)
was unclear.

This change adds three APIs in an "unstable" state:
- IEEE 802.15.4 L2 for subsystem developers
- IEEE 802.15.4 drivers for driver developers
- IEEE 802.15.4 net_mgmt for application developers

The corresponding APIs are documented at a group level.

These sub-APIs need to be separately versioned as they have different
audiences, change velocities and levels of stability.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: openthread: integrate ot/ieee802154 docs

Integrates and cross-references OpenThread and native IEEE 802.15.4 docs
as they share a common driver layer.

While OpenThread is of course not an IEEE 802.15.4 subsystem API it is
still included in the ieee802154 documentation group, purely for
convenience to the documentation readers.

For organizational purposes OT is considered a separate area of
maintenance and therefore keeps its dedicated documentation page - but
the two subsystems now cross-reference each other for convenience.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: connectivity/networking/utils: minor fixes

Fixes bugs in USB and DSA group titles (networking) and a minor bug in
sys/utils.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: ieee802154: mgmt: improved docs

Hides types in the IEEE 802.15.4 net_mgmt sub-API that are of no public
interest and improves documentation of the remainder.

All constants that are not meant to be used by applications have been
hidden. Only the API actually being of relevance to application
developers is now publicly visible and documented.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: ieee802154: l2: improved docs

Hides types used in the IEEE 802.15.4 L2 sub-API and L2-internal APIs
that are of no public interest and improves documentation of the
remainder.

The changes are mostly minor as the API documentation had already been
improved and clarified in previous changes. Also includes non-visible
documentation to the subsystem-internal Frame API by adding references
to the specification.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: ieee802154: driver: improved docs

Hides types in the IEEE 802.15.4 driver sub-API that are of no public
interest and improves documentation of the remainder.

Apart from extending and clarifying many details of the driver API, this
change also specifies certain formal properties of API calls, such as
"sleeps" or "isr-ok". Plus it removes semantic ambivalence that had been
resolved and hidden in existing implementation details but is highly
relevant to the way the API can be used and shall be implemented.

As far as possible this change introduces formal requirements language
(MAY, SHALL/MUST, SHOULD) to specify driver implementation requirements.

This change also introduces two definitions - for the moment being of
purely documentary relevance (i.e. to be referred in the documentation):
* IEEE802154_CONFIG_RX_SLOT_NONE
* IEEE802154_CONFIG_RX_SLOT_OFF

These "magic" values are required to clarify the exact semantics of RX
slots as to receiver activity.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

kartben commented:
> One thing that works well in Doxygen but not in the Sphinx version of the API is member grouping. See the undocumented driver API constants in Sphinx which are correctly documented in Doxygen via member groups. Not sure what's the problem there...

which ones?

rlubos commented on include/zephyr/net/ieee802154_radio.h:
```diff
@@ -779,12 +1510,27 @@ struct ieee802154_radio_api {
+	 * @returns a pointer to the driver's underlying network uptime
+	 * reference or NULL if not supported.
+	 */
+	const struct net_time_reference_api *(*get_time_reference)(const struct device *dev);
```
Any reference for `struct net_time_reference_api`? Is it proposed in some other PR?

fgrandel replied:
OMG, this should not have slipped in here. That comes from a downstream commit. Thx for catching this!

fgrandel commented on doc/connectivity/networking/api/ieee802154.rst:
```diff
@@ -7,30 +7,92 @@ IEEE 802.15.4
+Zephyr supports both, native IEEE 802.15.4 and Thread, with 6LoWPAN. Zephyr's
+:ref:`thread_protocol_interface` implementation is based on `OpenThread
+<https://openthread.io/>`_. The IPv6 header compression in 6LoWPAN is shared
+among native IEEE 802.15.4, OpenThread and the Bluetooth IPSP (IP support
```
Fixed.

fgrandel commented on include/zephyr/net/ieee802154_mgmt.h:
```diff
@@ -44,176 +50,245 @@ extern "C" {
+/**
+ * @name Event Macros
+ *
+ * @brief IEEE 802.15.4 net management commands.
```
Fixed.

fgrandel reviewed (commented):
@rlubos Thanks for the review. :-)

fgrandel commented:
@kartben @carlescufi Open for review now - proposes IEEE 802.15.4 as "unstable" and divided up as discussed.

@rlubos No changes introduced since your last review (except for improved commit messages). Needs re-approval, though, as I removed the draft status and rebased.

@jukkar Would be nice to hear your opinion as well re the proposal to split up into three sub-APIs and mark them as "unstable" for now.

kartben commented on include/zephyr/net/ieee802154.h:
```diff
@@ -23,44 +25,183 @@ extern "C" {
 
+/** interface-level security attributes, see section 9.5. */
 struct ieee802154_security_ctx {
+	/** section 9.5, secFrameCounter */
```
Picking just one example but this applies to many similar occurrences: is this really how you want to document these? I guess the intention is to not duplicate any information that could be found in the cited reference, but that makes for an odd/unhelpful "brief" description of the symbol, no? Shouldn't it be along the lines of "Frame counter bla bla bla.\n Refer to section 9.5, secFrameCounter for additional information."

fgrandel replied:
Very good question.

The only thing we've explicitly decided to be mandatory so far (#61227) is the spec pointer, so that we prove that we rely on standard primitives everywhere. The pointer definitely has practical relevance because it's the hard part to find the right section in the spec. I use these pointers myself all the time.

Additionally copying over at least a one-sentence definition from the spec in general might add value, though. In some instances I do it where I cannot assume that a typical developer will know the definition (and that was more or less my criterium so far). In most instances I don't when I just point to well known concepts that people in the subsystem should have a notion of.

The frame counter is a good example to discuss this. I'd expect any developer in the subsystem to know what it is. But should we copy over at least some single-sentence definition everywhere? I honestly don't have a strong opinion on that now.

I'd be very much interested in what you @kartben and @rlubos / @jukkar actually prefer as I'm definitely open to whatever you feel is the best solution.

jukkar replied:
IMHO I feel that the generated output might look a bit odd if we just have short ref to the spec. So I am inclined to agree with @kartben here. I am sure that for the expert the variable like frame_counter is obvious, I am just looking the generated doc and currently it looks a bit lonely there. See it yourself at
https://builds.zephyrproject.io/zephyr/pr/62503/docs/connectivity/networking/api/ieee802154.html#c.ieee802154_security_ctx.frame_counter

fgrandel replied:
@kartben @jukkar I wanted to add it to this PR but I underestimated Zephyr's speed once again. :-D I'll provide a one sentence definitions to all fields and concepts over the upcoming PRs. Thanks for your feedback there.

kartben replied:
mmh yeah why was this already merged? I guess we're missing something in GitHub UI to make it clear how long a PR has *actually*  been opened for review, since it hasn't even been 24 hours AFAICT.

fgrandel replied:
Guess it has to do with the closing 3.5 merge window? Or maybe just for the simple reason that the PR had two approvals and no request asked for officially. :-) Most certainly no deeper reason. In any case your comment is 100% valid has been confirmed by others which means I'll definitely take care of this and it will happen.

carlescufi reviewed (approved):
OK for the API maturity level changes
